runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,226 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * Persisted Queue supported by CorfuDB using distributed State Machine Replication.
+ * Entries enqueued are backed by a CorfuTable with a LinkedHashMap where each entry
+ * is mapped to a unique generated id that is returned upon successful <b>enqueue()</b>.
+ * <b>entryList()</b> returns the enqueued entries along with their special ids which can
+ * not only identify the element for removal but also has a global ordering defined by
+ * the order in which their <b>enqueue()</b> operations materialized.
+ * Instead of a dequeue() this Queue supports a <b>remove()</b> which accepts the id of the element.
+ * Entries cannot be modified in-place (or will lose ordering) but can be removed from anywhere
+ * from the persisted queue.
+ *
+ * Created by hisundar on 5/8/19.
+ *
+ * @param <E>   Type of the entry to be enqueued into the persisted queue
+ */
+@Slf4j
+public class CorfuQueue<E> {
+    /**
+     * The main CorfuTable which contains the primary key-value mappings.
+     */
+    private final CorfuTable<Long, E> corfuTable;
+    private final CorfuRuntime runtime;
+    private final CorfuGuidGenerator guidGenerator;
+
+    public CorfuQueue(CorfuRuntime runtime, String streamName) {
+        this.runtime = runtime;
+        corfuTable = runtime.getObjectsView().build()
+                .setTypeToken(new TypeToken<CorfuTable<Long, E>>() {})
+                .setStreamName(streamName)
+                .setArguments(CorfuTable.IndexRegistry.empty(), new LinkedHashMap<Long, E>())
+                .open();
+        guidGenerator = new CorfuGuidGenerator(runtime);
+    }
+
+    /**
+     * Returns the size of the queue at a given point in time.
+     */
+    public int size() {
+        return corfuTable.size();
+    }
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
+     * unless a transaction is used.
+     * Capacity restrictions and backoffs must be implemented outside this
+     * interface. Consider validating the size of the queue against a high
+     * watermark before enqueue.
+     *
+     * @param e the element to add
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this queue
+     * @return Globally ordered and unique ID representing this entry in the persistent queue
+     * WARNING: The ID returned by this function will NOT be the comparable with the ID
+     *          returned by the CorfuQueueRecord from entryList() method because if this
+     *          method executes within a transaction, the comparable ordering is not
+     *          known until the transaction commits.
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public UUID enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
```

Comment:
can we create a Dto class which would contain a UUID and a Long or something like that?
Using longs and uuids confusing

Reply:
That's what the first version had (CorfuQueueId), but I was asked to use UUID since consumers are using that. Since both you and @Maithem have asked for a separate type, I don't mind putting CorfuQueueId back, but I don't know the repercussions.

Reply:
To me - it had better use that class, just mark old API as deprecated. This mess with longs and UUIDs confusing

Reply:
done. using a custom class.